proposals/store_kyverno_policies_in_oci_registries.md

```diff
@@ -0,0 +1,129 @@
+
+The most helpful advantage of having an OCI Artifacts definition is that the community can continue to innovate, focusing on new artifact types without building another storage solution (YASS). More technically, OCI Artifacts are not a new specification, format, or API, but rather a set of conventions for storing assets other than images inside an OCI registry. For example, thanks to OCI Artifacts spec, additional artifacts, such as Helm Charts, can be easily stored and accessible from an OCI registry.
+
+# Definitions
+
+[definitions]: #definitions
+
+Make a list of the definitions that may be useful for those reviewing. Include phrases and words that Kyverno users or other interested parties may not be familiar with.
+
+# Motivation
+
+[motivation]: #motivation
+
+OCI Artifacts provides a [reference guide](https://github.com/opencontainers/artifacts/blob/main/artifact-authors.md) for authors looking to introduce a new artifact type. A unique artifact type is similar to defining a file extension. Defining a unique artifact allows various tools to work with the type uniquely. For example, it will enable a registry to display the type and tooling, such as vulnerability scanners, to know if and how they should interact with the contents. To learn more about defining your artifact type, please [see](https://github.com/opencontainers/artifacts/blob/main/artifact-authors.md#defining-a-unique-artifact-type).
+
+     “application/vnd.[org/company].[objectType].[optional sub type].config.[version]+json”
+
+# Proposal
+
+> Issue [#3154](https://github.com/kyverno/kyverno/issues/3154)
+
+Two implementation points exist in the Kyverno space that can be used to support pulling/pushing policies based on OCI Artifacts specification: Kyverno itself and its Command Line Interface (CL)I. First, we will start by introducing new subcommands for pulling/pushing policies using Kyverno CLI and then to Kyverno itself.
+
+One of the initial steps that OCI artifact authors need to decide upon is the mediaTypes that we will use to store Kyverno policies. According to the OCI Artifact Authors specification format defined for new artifact types, for the Kyverno policy layer, which we use to keep the policy itself, it should be represented as application/vnd.cncf.kyverno.policy.layer.v1+yaml, and application/vnd.cncf.kyverno.config.v1+json which will be used to store the policy config.
+
+The next step would be determining the library that we will use to interact with OCI Artifacts. The two well-known libraries that can help us achieve this are [go-containerregistry](https://github.com/google/go-containerregistry/) and [ORAS](https://oras.land). We can leverage either of them to pull and push Kyverno policies as OCI Artifacts.
+
+To push a Kyverno policy to a registry, the command would be:
+
+```shell
+kyverno push –policy <policy> <img>
+```
+
+To pull a policy from a registry, the command would be:
+
+```shell
+kyverno pull <img> –o <dir>
```

Comment:
Instead of doing a pull via a CLI, why not allow policies and other resources to be retrieved and applied in-memory from an OCI registry?

We can manage which policies should be loaded via a new resource:

```yaml
apiVersion: kyverno.io/v1
kind: PolicySets
spec:
  - name: pod-security
    oci: ghcr.io/myorg/psp:1.0
  - name: best-practices
    oci: ghcr.io/myorg/best-practices:2.1
  - name: multi-tenancy
    oci: ghcr.io/myorg/multi-tenancy:1.0
    policies:
      - add-network-policy
      - add-quota
```

Reply:
Awesome ideas🤩 but this KDP's context is a bit limited to kyverno CLI, I'd better discuss this in another KDP because I thought that it is nice to experience OCI Artifacts in Kyverno CLI first, then apply the same pattern to Kyverno's core, and once we get there maybe we could open another KDP to further discuss how it can be usable for Kyverno's core, WDYT?

Reply:
@developer-guy - we can certainly phase this out. 

However, what in your opinion would be the use case / benefit for users to manage policies in an OCI registry via the CLI? What I was thinking is that the main value of storing in an OCI registry is to make the policies easier to apply. Perhaps there are other use cases that you had in mind?

As part of this, will we update the CLI `apply` and `test` commands to take an OCI reference instead of a Git URL?

Reply:
>will we update the CLI apply and test commands to take an OCI reference instead of a Git URL or folder?

Exactly, we should expand the tests to use OCI references as an input in addition to Git URL or folder, does that make sense?

> However, what in your opinion would be the use case / benefit for users to manage policies in an OCI registry via the CLI? What I was thinking is that the main value of storing in an OCI registry is to make the policies easier to apply. Perhaps there are other use cases that you had in mind?

My answer would be these two things: `portability` and `security`.

> From the security perspective:

If I let people store Kyverno policies in OCI registries as OCI images, people can use existing Sigstore tooling to prove their authenticity by signing them, and we can verify that signature before applying them as the Flux team planned.

> From the portability perspective:
People can quickly share policies with others or copy them between registries.